lua/packer.lua

```diff
@@ -549,21 +549,32 @@ packer.compile = function(raw_args)
   manage_all_plugins()
   local args = parse_args(raw_args)
   local output_path = args.output_path or config.compile_path
+  local output_lua = vim.fn.fnamemodify(output_path, ':e') ~= 'vim'
   local should_profile = args.profile
   -- the user might explicitly choose for this value to be false in which case
   -- an or operator will not work
   if should_profile == nil then should_profile = config.profile.enable end
   refresh_configs(plugins)
   -- NOTE: we copy the plugins table so the in memory value is not mutated during compilation
-  local compiled_loader = compile(vim.deepcopy(plugins), should_profile)
+  local compiled_loader = compile(vim.deepcopy(plugins), output_lua, should_profile)
   output_path = vim.fn.expand(output_path)
   vim.fn.mkdir(vim.fn.fnamemodify(output_path, ":h"), 'p')
   local output_file = io.open(output_path, 'w')
   output_file:write(compiled_loader)
   output_file:close()
-  if config.auto_reload_compiled then vim.cmd("source " .. output_path) end
+  if config.auto_reload_compiled then
+    vim.cmd('source ' .. output_path)
+  end
   log.info('Finished compiling lazy-loaders!')
   packer.on_compile_done()
+
+  -- TODO: remove this after migration period (written 2021/06/28)
+  if output_lua then
+    local old_output_path = vim.fn.fnamemodify(output_path, ':r')..'.vim'
```

Comment:
I didn't see that the `packer_compiled` file registers its current location on startup (please correct me if I am wrong), so I decided to check for an old `packer_compiled` file by removing the `output_path`'s extension and appending '.vim' to it. This has a gotcha (see the commit message) but will handle the primary case of migrating users attempting to compile from one filetype to another in the same directory.

If the `packer_compiled` does log its location on startup, I can reference that here instead.

Reply:
Honestly, I never saw the rationale for auto-deleting "inconsistent" `packer_compiled` -- this is early access; just put a big fat warning somewhere that if anything looks fishy, manually delete the `packer_compiled` file and try again. It'll cease to be a problem within days, anyway.

Reply:
@clason: That's an option; the primary rationale for auto-deleting (from my perspective) is that having multiple `packer_compiled` files which both get sourced has caused a large number of annoying-to-debug problems for users in the past (e.g. issues where the user's spec and compiled file look fine and don't explain a behavior caused by an old compiled file they didn't realize was present). Adding a warning would help with this, but auto-deleting seems cleaner to me.

Reply:
Correct me if I'm wrong, but I think the one who implemented the lua runtime made it so only one of them would be sourced if there were both a lua file and a vim file

Reply:
It's more convenient for sure; I was just thinking since this is likely a one-time scenario (in a pre-release software, if you can get it in this week ;)), that seemed a bit like over-engineering to me.

Reply:
It's a fair point, yeah.

Reply:
I think it's worth it, given the aforementioned possibility of an obscure problem occuring for users. However, I think we should raise a warning and tell users to update their Neovim if they haven't already.

That way, we hopefully reduce the number of issue reports for the `packer_compiled` not being sourced and for the `packer_compiled` sourcing twice.